### PR TITLE
Separate Java and Scala typed TestProbes

### LINKED ISSUE #24542: Separate Testkit APIs for Java and Scala
Two separate interfaces with a shared internal impl.

Currently for example the Java TestProbe inherits the Scala one etc.

### PR DESCRIPTION
**Pull request by johanandren**
Fixes #24542  and #24334

### CODE REVIEW

**Comment by chbatey** [comment]
Do you plan to move the testkit as well as the test probe here?

The current Testkit doesn't really work with JUnit as @AfterClass needs to be static for tearing down the actor system.

**Comment by johanandren** [comment]
I'm not sure because just the testprobe became so many changes.

I wonder if we could do the Java testkit lifecycle as a `org.junit.rules.ExternalResource`, tying it to JUnit is maybe not optimal but it is so dominant in Java land that it may be ok?

**Comment by johanandren** [comment]
Let's make this about the TestProbe only and do the next step of separation in a separate PR for #24551

**Comment by johanandren** [comment]
Which means, ready for review.

File: akka-testkit-typed/src/main/scala/akka/testkit/typed/TestProbeImpl.scala [hunk 1]
```diff
@@ -0,0 +1,262 @@
...
+object FishingOutcome {
+
+  case object Continue extends FishingOutcome
+  case object ContinueAndIgnore extends FishingOutcome
+  case object Complete extends FishingOutcome
+  case class Fail(error: String) extends FishingOutcome
+}
+
+@InternalApi
+private[akka] object TestProbeImpl {
```
Comment (patriknw) [inline akka-testkit-typed/src/main/scala/akka/testkit/typed/TestProbeImpl.scala]: move to internal package?

File: akka-testkit-typed/src/main/scala/akka/testkit/typed/javadsl/TestProbe.scala [hunk 2]
```diff
@@ -3,42 +3,203 @@
...
 }
 
 /**
- * Java API:
+ * Java API: * Create instances through the `create` factories in the [[TestProbe]] companion.
+ *
+ * A test probe is essentially a queryable mailbox which can be used in place of an actor and the received
+ * messages can then be asserted etc.
+ *
+ * Not for user extension
```
Comment (patriknw) [inline akka-testkit-typed/src/main/scala/akka/testkit/typed/javadsl/TestProbe.scala]: add the annotation also

File: akka-testkit-typed/src/main/scala/akka/testkit/typed/scaladsl/TestProbe.scala [hunk 3]
```diff
@@ -288,56 +168,22 @@ class TestProbe[M](name: String)(implicit system: ActorSystem[_]) {
...
-              if (timeout.isFinite()) timeout - (System.nanoTime() - start).nanos
-              else timeout
-            if (newTimeout.toMillis <= 0) {
-              throw new AssertionError(s"timeout ($max) during fishForMessage, seen messages ${seen.reverse}, hint: $hint")
-            } else {
-              continue match {
-                case FishingOutcomes.Continue          ⇒ loop(newTimeout, msg :: seen)
-                case FishingOutcomes.ContinueAndIgnore ⇒ loop(newTimeout, seen)
-              }
+  def fishForMessage(max: FiniteDuration, hint: String)(fisher: M ⇒ FishingOutcome): List[M] =
```
Comment (patriknw) [inline akka-testkit-typed/src/main/scala/akka/testkit/typed/scaladsl/TestProbe.scala]: shouldn't we use `immutable.Seq` in public api?

**Review by patriknw — COMMENTED** [review]
looking good, some minors...

File: akka-testkit-typed/src/main/scala/akka/testkit/typed/FishingOutcome.scala [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+ * and [[akka.testkit.typed.javadsl.FishingOutcomes]]
+ */
+@DoNotInherit sealed trait FishingOutcome
+
+object FishingOutcome {
+
+  case object Continue extends FishingOutcome
+  case object ContinueAndIgnore extends FishingOutcome
+  case object Complete extends FishingOutcome
+  case class Fail(error: String) extends FishingOutcome
```
Comment (patriknw) [inline akka-testkit-typed/src/main/scala/akka/testkit/typed/FishingOutcome.scala]: `case class` should "always" be `final`

**Review by patriknw — APPROVED** [review]
LGTM

File: akka-testkit-typed/src/main/scala/akka/testkit/typed/FishingOutcome.scala (hunk 4 above)
Comment (johanandren) [reply akka-testkit-typed/src/main/scala/akka/testkit/typed/FishingOutcome.scala]: Fixed in the "other" PR

File: akka-actor-typed-tests/src/test/java/akka/actor/typed/ManualTimerTest.java [hunk 5]
```diff
@@ -14,14 +14,14 @@
...
-import akka.testkit.typed.ExplicitlyTriggeredScheduler;
+import akka.testkit.typed.javadsl.ExplicitlyTriggeredScheduler;
 import akka.testkit.typed.javadsl.TestProbe;
 
 public class ManualTimerTest extends TestKit {
   ExplicitlyTriggeredScheduler scheduler;
 
   public ManualTimerTest() {
-    super(parseString("akka.scheduler.implementation = \"akka.testkit.typed.ExplicitlyTriggeredScheduler\""));
+    super(parseString("akka.scheduler.implementation = \"akka.testkit.typed.javadsl.ExplicitlyTriggeredScheduler\""));
```
Comment (raboof) [inline akka-actor-typed-tests/src/test/java/akka/actor/typed/ManualTimerTest.java]: Should we have one implementation that implements both the javadsl and scaladsl interfaces here? kind of weird to refer to java/scala from the configuration.

File: akka-actor-typed-tests/src/test/java/akka/actor/typed/ManualTimerTest.java [hunk 6]
```diff
@@ -14,14 +14,14 @@
 import org.junit.Test;
 
 import akka.testkit.typed.TestKit;
-import akka.testkit.typed.ExplicitlyTriggeredScheduler;
+import akka.testkit.typed.javadsl.ExplicitlyTriggeredScheduler;
```
Comment (raboof) [inline akka-actor-typed-tests/src/test/java/akka/actor/typed/ManualTimerTest.java]: :+1:

File: akka-testkit-typed/src/main/scala/akka/testkit/typed/FishingOutcome.scala [hunk 7]
```diff
@@ -0,0 +1,22 @@
...
+
+import akka.annotation.DoNotInherit
+
+/**
+ * Not for user extension.
+ *
+ * Instances are available from `FishingOutcomes` in the respective dsls: [[akka.testkit.typed.scaladsl.FishingOutcomes]]
+ * and [[akka.testkit.typed.javadsl.FishingOutcomes]]
+ */
+@DoNotInherit sealed trait FishingOutcome
```
Comment (raboof) [inline akka-testkit-typed/src/main/scala/akka/testkit/typed/FishingOutcome.scala]: Do we need `DoNotInherit` if `FishingOutcome` is sealed? For Java users perhaps?

**Review by raboof — APPROVED** [review]
Important things look good, some minor comments

File: akka-testkit-typed/src/main/scala/akka/testkit/typed/FishingOutcome.scala (hunk 7 above)
Comment (johanandren) [reply akka-testkit-typed/src/main/scala/akka/testkit/typed/FishingOutcome.scala]: Yep, this one is for all the Java people out there!

File: akka-actor-typed-tests/src/test/java/akka/actor/typed/ManualTimerTest.java (hunk 5 above)
Comment (johanandren) [reply akka-actor-typed-tests/src/test/java/akka/actor/typed/ManualTimerTest.java]: I have more changes around this in the other PR, so let's discuss there.

**Pull request merged**